File: java/src/main/java/S031.java
public class S031 {

    private static final int[] DENOMS = new int[] {1, 2, 5, 10, 20, 50, 100, 200};
    private static final int TARGET = 200;

    public static int solve() {
        int[][] dp = new int[DENOMS.length + 1][TARGET + 1];
        // column 0 => make 0p => exactly 1 way to do this
        for (int r = 0; r <= DENOMS.length; r++) {
            dp[r][0] = 1;
        }
        // making >= 1p with no coins is impossible
        for (int c = 1; c <= TARGET; c++) {
            // extraneous, visual purposes
            dp[0][c] = 0;
        }
        for (int denomIdx = 1; denomIdx <= DENOMS.length; denomIdx++) {
            for (int curTarget = 1; curTarget <= TARGET; curTarget++) {
                int curDenom = DENOMS[denomIdx - 1];
                if (curDenom > curTarget) {
                    dp[denomIdx][curTarget] = dp[denomIdx - 1][curTarget];
                } else {
                    int ways = dp[denomIdx - 1][curTarget] + dp[denomIdx][curTarget - curDenom];
                    dp[denomIdx][curTarget] = ways;
                }
            }
        }
        return dp[DENOMS.length][TARGET];
    }

}

File: cpp/S031.cpp
#include <iostream>
#include <vector>

using namespace std;

static const vector<int> DENOMS = {1, 2, 5, 10, 20, 50, 100, 200};
static const int TARGET = 200;

int main() {
    // bottom-up DP using rolling tables
    // val @ row[i] = # ways to add to i for current number of coins
    vector<int> row(TARGET + 1);
    // there's 1 way to sum to 0
    row[0] = 1;
    for (int i = 0; i < DENOMS.size(); i++) {
        int cur_denom = DENOMS[i];
        // only consider totals greater than the current denom
        for (int t = cur_denom; t <= TARGET; t++) {
            row[t] += row[t - cur_denom];
        }
    }
    cout << row[TARGET] << endl;
    return 0;
}
